### ISSUE TITLE
Not all direct method invocations reach the module

### ISSUE THREAD

**Issue by GufCab** [Azure/iotedge#1555]
Only some Direct Method Invocations go through to edge module, about half never reach the module code. Only happens on one specific device so far, but very consistently. 

We have a test setup invoking a method every 5 minutes. Unlike issue #1550 the device starts responding again on its own - with no modules being restarted.

EdgeAgent, when started with debug, swamps the logs with debug messages, see below. This may be a symptom of the problem.

D2C messages go through without issues. Network connection seems quite stable when connecting through other applications (SSH, TeamViewer).

## Expected Behavior

When IoT Edge service is running, direct method invocations should go through. edgeAgent and edgeHub sends out log messages many times every second.

## Current Behavior

IoT edge service is running - about half of direct method invocations do not reach our custom module.

## Steps to Reproduce

1. Setup custom module to subscribe to direct method
2. Send direct method to module
3. About half will not reach the module

### Output of `iotedge check`

<details>
<summary>Click here</summary>

```
Configuration checks
--------------------
√ config.yaml is well-formed - OK
√ config.yaml has well-formed connection string - OK
√ container engine is installed and functional - OK
‼ config.yaml has correct hostname - Warning
    config.yaml has hostname skiolt-IOT-20190218-3 which does not comply with RFC 1035.

    - Hostname must be between 1 and 255 octets inclusive.
    - Each label in the hostname (component separated by ".") must be between 1 and 63 octets inclusive.
    - Each label must start with an ASCII alphabet character (a-z), end with an ASCII alphanumeric character (a-z, 0-9), and must contain only ASCII alphanumeric characters or hyphens (a-z, 0-9, "-").

    Not complying with RFC 1035 may cause errors during the TLS handshake with modules and downstream devices.
√ config.yaml has correct URIs for daemon mgmt endpoint - OK
√ latest security daemon - OK
√ host time is close to real time - OK
√ container time is close to host time - OK
‼ DNS server - Warning
    Container engine is not configured with DNS server setting, which may impact connectivity to IoT Hub.
    Please see https://aka.ms/iotedge-prod-checklist-dns for best practices.
    You can ignore this warning if you are setting DNS server per module in the Edge deployment.
‼ production readiness: certificates - Warning
    Device is using self-signed, automatically generated certs.
    Please see https://aka.ms/iotedge-prod-checklist-certs for best practices.
√ production readiness: certificates expiry - OK
√ production readiness: container engine - OK
‼ production readiness: logs policy - Warning
    Container engine is not configured to rotate module logs which may cause it run out of disk space.
    Please see https://aka.ms/iotedge-prod-checklist-logs for best practices.
    You can ignore this warning if you are setting log policy per module in the Edge deployment.

Connectivity checks
-------------------
√ host can connect to and perform TLS handshake with IoT Hub AMQP port - OK
√ host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - OK
√ host can connect to and perform TLS handshake with IoT Hub MQTT port - OK
√ container on the default network can connect to IoT Hub AMQP port - OK
√ container on the default network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the default network can connect to IoT Hub MQTT port - OK
√ container on the IoT Edge module network can connect to IoT Hub AMQP port - OK
√ container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the IoT Edge module network can connect to IoT Hub MQTT port - OK
√ Edge Hub can bind to ports on host - OK

19 check(s) succeeded.
```

</details>

### Device (Host) Operating System
Ubuntu 18.04.3 LTS

### Architecture

x86_64

### Container Operating System

Linux containers

#### iotedged

iotedge 1.0.8 

#### Edge Agent

1.0.8 (and 1.0.7)
Does not happen on 1.0.6

#### Edge Hub

1.0.8 (and 1.0.7)
Does not happen on 1.0.6

#### Docker

```
Client:
 Version:           3.0.6
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        a63faebc
 Built:             Wed Jun 19 22:56:37 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          3.0.6
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       b07f53d
  Built:            Wed Jun 19 22:28:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

## Logs

<details>
<summary>edge-agent logs</summary>

The edge agent log is filled with this

```
2019-08-09 14:56:40.980 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:40.980 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:40.980 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.011 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.011 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.011 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.041 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.041 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.042 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.118 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.118 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.118 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.149 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.149 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.149 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.182 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.182 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.182 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.214 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.214 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.214 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.246 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.246 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.246 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...
2019-08-09 14:56:41.335 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Edge agent updated deployment config from desired properties.
2019-08-09 14:56:41.335 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Updated edge agent configuration from upstream twin.
2019-08-09 14:56:41.335 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection] - Begin refreshing twin from upstream...

```
</details>

<details>
<summary>edge-hub logs</summary>

```
2019-08-09 15:02:46.674 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:46.706 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:46.736 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:46.828 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:46.848 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:46.918 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:46.929 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:46.981 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:46.995 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.040 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:47.120 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.143 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:47.180 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.181 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.202 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:47.294 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.299 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:47.307 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.366 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:47.388 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.420 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.485 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:47.489 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.559 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:47.628 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.661 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:47.695 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.743 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:47.810 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.868 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/distriwin-proxy
2019-08-09 15:02:47.869 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub
2019-08-09 15:02:47.926 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Connected and reason Connection_Ok for 208000000000001/$edgeHub
2019-08-09 15:02:48.002 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/distriwin-proxy
2019-08-09 15:02:48.033 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected_Retrying and reason No_Network for 208000000000001/$edgeHub

```
</details>

## Additional Information

This may be an issue on the device itself manifesting in the edgeAgent and hub behaving strangely. 
Works if we revert to using edgeAgent and edgeHub versions 1.0.6.

Removing custom module so only edgeAgent end edgeHub are running on the device leads to the same logs.

**Comment by varunpuranik** [comment]
@GufCab - We made a fix related to subscriptions in 1.0.8.1. Can you try to update your EdgeHub module to 1.0.8.1 and see if the issue still reproduces? If it does, then can you provide the full debug logs for the EdgeHub module? That will help us investigate what might be going on.

**Comment by lt72** [comment]
Closing this issue as it seems stale, please feel free to re-open if you need to.
